libs/core/langchain_core/indexing/base.py

```diff
@@ -317,3 +317,16 @@ def delete_keys(self, keys: Sequence[str]) -> None:
 
     async def adelete_keys(self, keys: Sequence[str]) -> None:
         self.delete_keys(keys)
+
+
+class UpsertResponse(TypedDict):
+    """A generic response for upsert operations.
+
+    The upsert response will be used by abstractions that implement an upsert
+    operation for content that can be upserted by ID.
+    """
+
+    succeeded: List[str]
+    """The IDs that were successfully indexed."""
+    failed: List[str]
```

Comment:
(nit) are error messages useful to capture here

Reply:
No, just the IDs right now. We can expand later with a NotRequired key for error message? I'm not sure what useful just yet -- (an alternative is standardizing on exceptions)